File: app/src/main/java/com/haishinkit/amf/AMF0Serializer.java
package com.haishinkit.amf;

import com.haishinkit.as3.ASUndefined;

import java.util.Map;
import java.util.List;
import java.util.Date;
import java.nio.ByteBuffer;

public final class AMF0Serializer {
    private final ByteBuffer buffer;

    public AMF0Serializer(final ByteBuffer buffer) {
        this.buffer = buffer;
    }

    public AMF0Serializer putBoolean(final boolean value) {
        buffer.put(AMF0Marker.BOOL.valueOf());
        buffer.put((byte)(value ? 1 : 0));
        return this;
    }

    public AMF0Serializer putDouble(final double value) {
        buffer.put(AMF0Marker.NUMBER.valueOf());
        buffer.putDouble(value);
        return this;
    }

    public AMF0Serializer putString(final String value) {
        if (value == null) {
            buffer.put(AMF0Marker.NULL.valueOf());
            return this;
        }
        int length = value.length();
        boolean isShort = length <= (int) Short.MAX_VALUE ? true : false;
        buffer.put(isShort ? AMF0Marker.STRING.valueOf() : AMF0Marker.LONGSTRING.valueOf());
        return putString(value, isShort);
    }

    public AMF0Serializer putMap(final Map<String, Object> value) {
        if (value == null) {
            buffer.put(AMF0Marker.NULL.valueOf());
            return this;
        }
        buffer.put(AMF0Marker.OBJECT.valueOf());
        for (Map.Entry<String, Object> entry: value.entrySet()) {
            putString(entry.getKey(), true).putObject(entry.getValue());
        }
        putString("", true);
        buffer.put(AMF0Marker.OBJECTEND.valueOf());
        return this;
    }

    public AMF0Serializer putDate(final Date value) {
        if (value == null) {
            buffer.put(AMF0Marker.NULL.valueOf());
            return this;
        }
        buffer.put(AMF0Marker.DATE.valueOf());
        buffer.putDouble(new Long(value.getTime()).doubleValue());
        buffer.put(new byte[]{0x00, 0x00});
        return this;
    }

    public AMF0Serializer putList(final List<Object> value) {
        if (value == null) {
            buffer.put(AMF0Marker.NULL.valueOf());
            return this;
        }
        buffer.put(AMF0Marker.ECMAARRAY.valueOf());
        if (value.isEmpty()) {
            buffer.put(new byte[]{0x00, 0x00, 0x00, 0x00});
            return this;
        }
        buffer.putInt(value.size());
        for (Object object : value) {
            putObject(object);
        }
        return this;
    }

    public AMF0Serializer putObject(final Object value) {
        if (value == null) {
            buffer.put(AMF0Marker.NULL.valueOf());
            return this;
        }
        if (value instanceof String) {
            return putString((String) value);
        }
        if (value instanceof Double) {
            return putDouble((Double) value);
        }
        if (value instanceof Integer) {
            return putDouble(((Integer) value).doubleValue());
        }
        if (value instanceof Boolean) {
            return putBoolean((Boolean) value);
        }
        if (value instanceof Date) {
            return putDate((Date) value);
        }
        if (value instanceof Map) {
            return putMap((Map<String, Object>) value);
        }
        if (value instanceof List) {
            return putList((List<Object>) value);
        }
        if (value instanceof ASUndefined) {
            buffer.put(AMF0Marker.UNDEFINED.valueOf());
            return this;
        }
        return this;
    }

    public String toString() {
        return buffer.toString();
    }

    private AMF0Serializer putString(final String value, final boolean asShort) {
        int length = value.length();
        if (asShort) {
            buffer.putShort((short) length);
        } else {
            buffer.putInt(length);
        }
        buffer.put(value.getBytes());
        return this;
    }
}

File: haishinkit/src/main/java/com/haishinkit/amf/Amf0Serializer.kt
package com.haishinkit.amf

import com.haishinkit.amf.data.AsUndefined
import java.nio.ByteBuffer
import java.util.Date

internal class Amf0Serializer(private val buffer: ByteBuffer) {
    fun putBoolean(value: Boolean): Amf0Serializer {
        buffer.put(Amf0Marker.BOOL.rawValue)
        buffer.put((if (value) 1 else 0).toByte())
        return this
    }

    fun putDouble(value: Double): Amf0Serializer {
        buffer.put(Amf0Marker.NUMBER.rawValue)
        buffer.putDouble(value)
        return this
    }

    fun putString(value: String?): Amf0Serializer {
        if (value == null) {
            buffer.put(Amf0Marker.NULL.rawValue)
            return this
        }
        val length = value.length
        val isShort = length <= java.lang.Short.MAX_VALUE.toInt()
        buffer.put(if (isShort) Amf0Marker.STRING.rawValue else Amf0Marker.LONGSTRING.rawValue)
        return putString(value, isShort)
    }

    fun putMap(value: Map<String, Any?>?): Amf0Serializer {
        if (value == null) {
            buffer.put(Amf0Marker.NULL.rawValue)
            return this
        }
        buffer.put(Amf0Marker.OBJECT.rawValue)
        for ((key, value1) in value) {
            putString(key, true).putObject(value1)
        }
        putString("", true)
        buffer.put(Amf0Marker.OBJECTEND.rawValue)
        return this
    }

    fun putDate(value: Date?): Amf0Serializer {
        if (value == null) {
            buffer.put(Amf0Marker.NULL.rawValue)
            return this
        }
        buffer.put(Amf0Marker.DATE.rawValue)
        buffer.putDouble(value.time.toDouble())
        buffer.put(byteArrayOf(0x00, 0x00))
        return this
    }

    fun putList(value: List<Any>?): Amf0Serializer {
        if (value == null) {
            buffer.put(Amf0Marker.NULL.rawValue)
            return this
        }
        buffer.put(Amf0Marker.ECMAARRAY.rawValue)
        if (value.isEmpty()) {
            buffer.put(byteArrayOf(0x00, 0x00, 0x00, 0x00))
            return this
        }
        buffer.putInt(value.size)
        for (`object` in value) {
            putObject(`object`)
        }
        return this
    }

    fun putObject(value: Any?): Amf0Serializer {
        if (value == null) {
            buffer.put(Amf0Marker.NULL.rawValue)
            return this
        }
        if (value is String) {
            return putString(value as String?)
        }
        if (value is Double) {
            return putDouble((value as Double?)!!)
        }
        if (value is Int) {
            return putDouble(value.toDouble())
        }
        if (value is Short) {
            return putDouble(value.toDouble())
        }
        if (value is Boolean) {
            return putBoolean((value as Boolean?)!!)
        }
        if (value is Date) {
            return putDate(value as Date?)
        }
        if (value is Map<*, *>) {
            return putMap(value as Map<String, Any>?)
        }
        if (value is List<*>) {
            return putList(value as List<Any>?)
        }
        if (value is AsUndefined) {
            buffer.put(Amf0Marker.UNDEFINED.rawValue)
            return this
        }
        return this
    }

    override fun toString(): String {
        return buffer.toString()
    }

    private fun putString(
        value: String,
        asShort: Boolean,
    ): Amf0Serializer {
        val length = value.length
        if (asShort) {
            buffer.putShort(length.toShort())
        } else {
            buffer.putInt(length)
        }
        buffer.put(value.toByteArray())
        return this
    }
}
